service/matching/taskQueueManager.go

```diff
@@ -748,92 +683,100 @@ func (c *taskQueueManagerImpl) TaskQueueKind() enumspb.TaskQueueKind {
-	// If it's nil because it never existed, then we'd never have any data.
-	// It can't be nil due to removing versions, as that would result in a non-nil container with
-	// nil inner fields.
-	if res.GetUserData() != nil {
-		c.db.setUserDataForNonRootPartition(res.GetUserData())
-	}
-	// We want to start the poller as long as the root partition has any kind of data (or fetching hasn't worked)
-	if res.GetUserData() != nil || err != nil {
-		c.metadataPoller.StartIfUnstarted()
-	}
+	fetchSource, err := c.userDataFetchSource()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return res.GetUserData(), nil
-}
 
-// StartIfUnstarted starts the poller if it's not already started. The passed in function is called repeatedly
-// and if it returns true, the poller will shut down, at which point it may be started again.
-func (mp *metadataPoller) StartIfUnstarted() {
-	if mp.running.Load() {
-		return
+	op := func(ctx context.Context) error {
+		knownUserData, _, err := c.db.GetUserData(ctx)
+		if err != nil && !errors.Is(err, errUserDataNotPresentOnPartition) {
+			return err
+		}
+
+		callCtx, cancel := context.WithTimeout(ctx, c.config.GetUserDataLongPollTimeout())
+		defer cancel()
+
+		res, err := c.matchingClient.GetTaskQueueUserData(callCtx, &matchingservice.GetTaskQueueUserDataRequest{
+			NamespaceId:              c.taskQueueID.namespaceID.String(),
+			TaskQueue:                fetchSource,
+			LastKnownUserDataVersion: knownUserData.GetVersion(),
+			WaitNewData:              true,
+		})
+		if err != nil {
+			return err
+		}
+		// If the root partition returns nil here, then that means our data matched, and we don't need to update.
+		// If it's nil because it never existed, then we'd never have any data.
+		// It can't be nil due to removing versions, as that would result in a non-nil container with
+		// nil inner fields.
+		if res.GetUserData() != nil {
+			c.db.setUserDataForNonOwningPartition(res.GetUserData())
+		}
+		resolveInitialFetch(nil)
+		return nil
 	}
-	go mp.pollLoop()
-}
 
-func (mp *metadataPoller) pollLoop() {
-	mp.running.Store(true)
-	defer mp.running.Store(false)
-	ticker := time.NewTicker(mp.pollIntervalCfgFn())
-	defer ticker.Stop()
+	minWaitTime := c.config.GetUserDataMinWaitTime
 
-	for {
-		select {
-		case <-mp.stopChan:
-			return
-		case <-ticker.C:
-			// In case the interval has changed
-			ticker.Reset(mp.pollIntervalCfgFn())
-			data, err := mp.tqMgr.fetchUserDataFromRootPartition(context.TODO())
-			if data == nil && err == nil {
-				// Can stop polling since there is no versioning data. Loop will be restarted if we
-				// are told to invalidate the data, or we attempt to fetch it via GetVersioningData.
-				return
+	for ctx.Err() == nil {
+		start := time.Now()
+		_ = backoff.ThrottleRetryContext(ctx, op, getUserDataRetryPolicy, nil)
+		elapsed := time.Since(start)
+
+		// In general we want to start a new call immediately on completion of the previous
```

Comment:
Can you explain why this protection is in place?
Why would the remote return success immediately?

Reply:
> But if the remote is broken

That's all, I just don't want a bug that causes the server to return success immediately to lead to this spinning. But nor do I want to delay the follow up call after a successful call if the call took 30s. It shouldn't happen (more than once) if things are working properly